File: morkato-bot/src/main/java/org/morkato/bot/extension/MorkatoAPIExtension.java
package org.morkato.bot.extension;

import jakarta.validation.Validation;
import jakarta.validation.Validator;
import org.flywaydb.core.Flyway;
import org.morkato.api.repository.guild.GuildRepository;
import org.morkato.boot.annotation.ApplicationProperty;
import org.morkato.boot.annotation.DefaultValue;
import org.morkato.boot.ApplicationContext;
import org.morkato.bmt.context.BotContext;
import org.morkato.boot.BaseExtension;
import org.morkato.database.PsqlRepositoryStatement;
import org.morkato.database.repository.guild.PsqlGuildRepository;
import org.morkato.jdbc.QueryLoader;
import org.morkato.jdbc.ReflectionQueryLoader;
import org.springframework.jdbc.datasource.DriverManagerDataSource;

import javax.sql.DataSource;

public class MorkatoAPIExtension extends BaseExtension<BotContext> {
  private final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();

  @ApplicationProperty("morkato.datasource.url")
  private String dbUrl;
  @ApplicationProperty("morkato.datasource.user")
  private String dbUser;
  @ApplicationProperty("morkato.datasource.password")
  private String dbPassword;
  @ApplicationProperty("morkato.datasource.driver")
  private String dbDriver;
  @ApplicationProperty("morkato.datasource.migrationPath")
  @DefaultValue("classpath:migration")
  private String migrationPath;
  @ApplicationProperty("morkato.datasource.queriesPath")
  @DefaultValue("queries")
  private String queriesPath;

  public DataSource connect() {
    final DriverManagerDataSource source = new DriverManagerDataSource();
    source.setDriverClassName(dbDriver);
    source.setUrl(dbUrl);
    source.setUsername(dbUser);
    source.setPassword(dbPassword);
    return source;
  }

  @Override
  public void start(ApplicationContext<BotContext> application) {
    final DataSource source = this.connect();
    final PsqlRepositoryStatement statement = new PsqlRepositoryStatement(source, validator);
    final QueryLoader loader = new QueryLoader(queriesPath);
    final ReflectionQueryLoader reflection = new ReflectionQueryLoader(loader);
    final Flyway flyway = Flyway.configure()
      .dataSource(source)
      .locations(migrationPath)
      .load();
    flyway.migrate();
    GuildRepository repository = new PsqlGuildRepository(statement);
    reflection.writeAll(repository);
    application.inject(repository);
  }
}

File: morkato-bot/src/main/kotlin/org/morkato/bot/extension/MorkatoAPIExtension.kt
package org.morkato.bot.extension

import jakarta.validation.Validation
import org.jetbrains.exposed.sql.Database
import org.morkato.api.repository.SimpleRepositoryCentral
import org.morkato.bmt.annotation.ApplicationProperty
import org.morkato.bmt.annotation.RegistryExtension
import org.morkato.bmt.context.ApplicationContextBuilder
import org.morkato.bmt.extensions.BaseExtension
import org.morkato.database.DatabaseProvider
import org.morkato.database.repository.DatabaseGuildRepository

@RegistryExtension
class MorkatoAPIExtension : BaseExtension() {
//  companion object {
//    val LOGGER: Logger = LoggerFactory.getLogger(MorkatoAPIExtension::class.java)
//    /* TODO: Está extensão será removida. Farei um banco virtual por agora, pois criar no físico (PostgreSQL) demanda mais tempo. Isso é temporário. */
//    val builtinTrainers: Set<TrainerCreationQuery> = setOf(
//      TrainerCreationQuery(
//        "971803172056219728",
//        "Novato",
//        131162,
//        BigDecimal.valueOf(1000),
//        BigDecimal.valueOf(1000),
//        BigDecimal.valueOf(0),
//        "⚔",
//        "https://c.tenor.com/fWzS8rODMGwAAAAd/tenor.gif",
//        0
//      )
//    )
//  }

  @ApplicationProperty("morkato.datasource.url")
  lateinit var dbUrl: String
  @ApplicationProperty("morkato.datasource.user")
  lateinit var dbUser: String
  @ApplicationProperty("morkato.datasource.password")
  lateinit var dbPassword: String

  override fun start(app: ApplicationContextBuilder) {
    val central = SimpleRepositoryCentral()
    app.inject(central)
    val validator = Validation.buildDefaultValidatorFactory().getValidator()
    val datasource = DatabaseProvider.getDatasource(app.getProperties())
    val flyway = DatabaseProvider.getFlyway(datasource)
    flyway.migrate()
    val database = Database.connect(datasource)
    central.setGuildRepository(DatabaseGuildRepository(database, validator))
  }
}